### ISSUE TITLE
OOM issue with same batch size that was running ok on 0.0.80

### ISSUE THREAD

**Issue by salrowili** [erfanzar/EasyDeL#184]
Hi,

I've noticed that recent updates are causing the SFT trainer code to throw an OutOfMemory (OOM) error with the same batch size that previously ran without issue on version 0.0.80.

I attempted SFT tuning using bfloat16 (no LoRA) with LLaMA 3.1 8B, max_length=1024, and batch=8 on TPUv4-8, but encountered an OOM error. This fine-tuning setup was working ok with 0.0.80.

**Comment by salrowili** [comment]
Hi @erfanzar ,
Please see the code below : 

```
import easydel as ed
from easydel.utils.analyze_memory import SMPMemoryMonitor  # Optional for memory analysis
import jax
from jax import numpy as jnp, sharding, lax, random as jrnd
from huggingface_hub import HfApi
import datasets
from flax.core import FrozenDict
from datasets import load_dataset
import jsonlines ## pip install jsonlines
import torch
import numpy as np
import pandas as pd
from datasets import load_dataset
from transformers import AutoTokenizer, AutoModelForCausalLM, TrainingArguments
import wandb
wandb.init(project="test")
def train():
	pretrained_model_name_or_path="meta-llama/Llama-3.1-8B-Instruct"
	max_length=1024
	PartitionSpec, api = sharding.PartitionSpec, HfApi()
	json_entry=[]
	qa_data = load_dataset("Stanford/web_questions" ,split="train")
	instruction ="You are a helpful AI assistant."
	for item in qa_data:
		question=item["question"]
		answer=item["answers"][0]
		json_entry.append({"messages": [{"role": "system", "content": instruction}, {"role": "user", "content": item["question"]}, {"role": "assistant", "content": answer}]})

	with jsonlines.open('SFT_Train.jsonl', 'w') as writer:
    		writer.write_all(json_entry)

	train_dataset = load_dataset("json", data_files={"train" :"/home/big35manf/SFT_Train.jsonl"},split="train")
	sharding_axis_dims = (1, 1, -1, 1)
	new_repo_id = "Test"
	dtype = jnp.bfloat16

	model = ed.AutoEasyDeLModelForCausalLM.from_pretrained(
		pretrained_model_name_or_path,
		auto_shard_model=True,
		sharding_axis_dims=sharding_axis_dims,
		config_kwargs=ed.EasyDeLBaseConfigDict(
			use_scan_mlp=False,
			attn_dtype=jnp.bfloat16,
			freq_max_position_embeddings=max_length,
			mask_max_position_embeddings=max_length,
			attn_mechanism=ed.AttentionMechanisms.VANILLA,
		),
		param_dtype=dtype,
		torch_dtype=torch.bfloat16,
		dtype=dtype,
        	from_torch=True,
		precision=lax.Precision("fastest"),
	)
	config = model.config
	tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path,trust_remote_code=True)
	tokenizer.pad_token = (tokenizer.eos_token if tokenizer.pad_token is None else tokenizer.pad_token)
	tokenizer.padding_side = "right"
	train_arguments = ed.SFTConfig(
		num_train_epochs=10,
		learning_rate=5e-5,
		learning_rate_end=0,
		warmup_steps=100,
		optimizer=ed.EasyDeLOptimizers.ADAMW,
		scheduler=ed.EasyDeLSchedulers.WARM_UP_COSINE,
		weight_decay=0.02,
		total_batch_size=8,
		max_sequence_length=max_length,
		gradient_accumulation_steps=1,
		do_last_save=True,
		model_name=new_repo_id,
		track_memory=False,
        	packing=True,
        	num_of_sequences=max_length,
		dataset_text_field=None,
		dataset_num_proc=32,
	)

	trainer = ed.SFTTrainer(
		processing_class=tokenizer,
		arguments=train_arguments,
		model=model,
		train_dataset=train_dataset,
		eval_dataset=None,
		formatting_func=lambda x: [tokenizer.apply_chat_template(x["messages"], tokenize=False)]
	)

	output = trainer.train()
	logger.info("Training Done")
	tokenizer.save_pretrained(output.last_save_file_name)
train()

```

**Comment by erfanzar** [comment]
Can you rerun the code? There was an issue with the loss function, which wasn't using the fused version.

**Comment by erfanzar** [comment]
And since the sharding mechanism you're using is tensor parallel you can except OOM but not on 1k sequence length

**Comment by erfanzar** [comment]
In v0.0.80 trainer will automatically use gradient checkpointing (this behavior is removed in 0.1.0 and you should pass gradient_checkponiting to model_kwargs (ill take blame for not having good documentation))

**Comment by salrowili** [comment]
You are right!. In 0.0.80, it was part of training arguments as we can see in this example : 
https://github.com/erfanzar/EasyDeL/blob/4c01dfe2335bd41ad375d0d112024153813621f7/docs/supervised-fine-tuning-trainer-example.md?plain=1#L119
However, it was removed in the recent updates
https://github.com/erfanzar/EasyDeL/blob/4c01dfe2335bd41ad375d0d112024153813621f7/easydel/trainers/training_configurations.py#L67

after updating the  code with : 

```
                config_kwargs=ed.EasyDeLBaseConfigDict(
                        use_scan_mlp=False,
                        attn_dtype=jnp.bfloat16,
                        freq_max_position_embeddings=max_length,
                        mask_max_position_embeddings=max_length,
                        attn_mechanism=ed.AttentionMechanisms.VANILLA,
                        gradient_checkpointing=ed.EasyDeLGradientCheckPointers.NOTHING_SAVEABLE,
                ),
```

i was able to run SFT code with 8 batch size but i got  couple of warning:

`[easydel.trainers.base_trainer] Prevent Running Model Due to NaN Loss`

`FutureWarning: unhashable type: <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>. Attempting to hash a tracer will lead to an error in a future JAX release.`

The issue of NaN is presented even with 0.0.80 every time i use meta-llama/Llama-3.1-8B-Instruct model with Packing=True. It would go in some runs and in some runs will appear. I have worked around this issue by re-running the script multiple time till i had no NaN, without changing any arguments. With Packing=False, the issue would disappear .This issue is not presented with other llama3.2 models.

**Comment by erfanzar** [comment]
Look flops calculation method is changed in last version every thing was manually calculated but in this version it's calculated from jax analysis so except it to be wrong for example you might be running and getting 160flops but in some parts xla play a bit dumb and show 130Flops

**Comment by erfanzar** [comment]
Thank you @salrowili for bringing up these issues and for your detailed feedback!
I want to clarify that several of the speed-related issues stem from our flax/NNX integration. I'm currently working on fixing these with @cgarciae's help. Going forward, any easydel issues that are specifically related to flax/NNX will be redirected to the flax/NNX repository with additional context and documentation.

**Comment by salrowili** [comment]
Great!. Thank you @erfanzar for opening the topic. I have one question. I am planning to start sharing my code with the topic you just opened https://github.com/erfanzar/EasyDeL/discussions/185, but i am still struggling to run my codes on the new EasyDEL 0.1dev release. Its very slow compared to 0.0.80 and you have told me that it due to flax/NNX integration. Do you have any estimation when the issue will be fixed? Because if it will be soon, i will wait and till it fixed and share my codes with 0.1dev release.

**Comment by erfanzar** [comment]
Hi @salrowili,  

Many performance issues related to the new arguments and the updated base trainer have been resolved. These include fixes for duplicated `if` statements, redundant code checks, and incorrect caching mechanisms.  

You can rerun your benchmark to see if there are any remaining performance issues (avoid using ahead-of-time compilation).  

With Qwen-2 7B, batch size 8, and full sequence parallelism, I was able to achieve **6 seconds per iteration**. Let me know how it goes!

**Comment by erfanzar** [comment]
@salrowili im using tests/trainer_test.py

**Comment by erfanzar** [comment]
Hi @salrowili,

I hope the performance issue is running smoothly now. If you’re still encountering any other problems, please let me know. I’m currently working on improving the speed of the training process and would be happy to assist you further.

For quicker communication and to resolve any issues more efficiently, feel free to connect with me on Discord. My user ID is `citifer`. Looking forward to helping you out!

**Comment by salrowili** [comment]
Hi @erfanzar . The issue is still there and you can verify it by using the the SFT code that i posted early.  Currently, i am doing the SFT training with 0.0.80 version. How i can disable AOT compilation?

I thought i should slow down in reporting issues since you may seems busy with other EasyDel issues, and this why i did not update you that the issue still exists (: . 

I will try to DM you on Discord soon

**Comment by salrowili** [comment]
0.0.80 is much faster especially when you run larger model. I ran llama3.1 8B and with 0.1dev it gave me 1.9s/it with ~2H against 1.24s/it with 1H 15m

**Comment by erfanzar** [comment]
Hi @salrowili,  

Could you please test and confirm whether the execution time and `step_times` in `easydel-v0.1.0-dev` are approximately 10–20% faster compared to `easydel-0.80.0`?  

Based on my tests using V3-8:  
- Version `0.80.0` runs at **2 seconds per iteration** with your provided code.  
- Version `0.1.0-dev` runs at **1.6 seconds per iteration** with the same config.  

Let me know your findings!

**Comment by salrowili** [comment]
Hi @erfanzar . I think i have figured out the root of the speed issue. I think you are testing the code on TPUv3-8 which has 8 chips each has single core. In contrast, TPUv4-8 has 4 chips each has 2 cores. Thus, maybe you are only using single core from each chip on TPUv4-8??.  This also may impact how you calculate FLOPs.

I having tested 0.1 on both TPUv3-8 and TPUv4-8 and TPUv3-8 is much faster (almost double). Also the trainer output is messed up as it print output vertically not horizontally as it used to.

Here is Trainer output for TPUv3-8 and TPUv4-8. The code for both is identical including the sharding method. Observe the step time.

TPUv3-8

```
training process - { 'mlperf/execution_time': 0.37421172,                                                                                                                      
  'mlperf/flops': 8245941370880.0,                                                                                                                                             
  'mlperf/flops_per_sequence': 4173047252.4696355,                                                                                                                             
  'mlperf/flops_per_token': 4075241.4574898784,                                                                                                                                
  'mlperf/flops_pre_second': 21753546211775.117,                                                                                                                               
  'mlperf/flops_sequence_pre_second': 4173047252.4696355,                                                                                                                      
  'mlperf/flops_token_pre_second': 4075241.4574898784,                                                                                                                         
  'mlperf/step_time': 0.3790619373321533,                                                                                                                                      
  'mlperf/throughput': 21611.24395040949,                                                                                                                                      
  'mlperf/total_time': 265.9115765094757,                                                                                                                                      
  'train/accuracy': 0.9125122427940369,                                                                                                                                        
  'train/epoch': 9,                                                                                                                                                            
  'train/grad_norm/lm_head.kernel': 0.35546875,                                                                                                                                
  'train/grad_norm/model.norm.kernel': 0.0031585693359375,                                                                                                                     
  'train/learning_rate': 9.386166652802785e-07,                                                                                                                                
  'train/loss': 0.3316769003868103,                                                                                                                                            
  'train/mean_accuracy': 0.8708844780921936,                                                                                                                                   
  'train/mean_loss': 0.6994993090629578,                                                                                                                                       
  'train/perplexity': 1.3933013677597046,                                                                                                                                      
  'train/step': 246,                                                                                                                                                           
  'train/train/max_grad_norm': 1.3671875,                                                                                                                                      
  'train/train/mean_grad_norm': 0.10693359375,                                                                                                                                 
  'train/visited_tokens': 2023424,                                                                                                                                             
  'train/z_loss': 0.0}  
```


TPUv4-8

```
training process - { 'mlperf/execution_time': 0.772636662,                                                                                                                     
  'mlperf/flops': 115442874056704.0,                                                                                                                                           
  'mlperf/flops_per_sequence': 53052791386.35294,                                                                                                                              
  'mlperf/flops_per_token': 51809366.5882353,                                                                                                                                  
  'mlperf/flops_pre_second': 148604274400124.53,                                                                                                                               
  'mlperf/flops_sequence_pre_second': 53052791386.35294,                                                                                                                       
  'mlperf/flops_token_pre_second': 51809366.5882353,                                                                                                                           
  'mlperf/step_time': 0.7768476009368896,                                                                                                                                      
  'mlperf/throughput': 10545.182851978081,                                                                                                                                     
  'mlperf/total_time': 479.89153575897217,                                                                                                                                     
  'train/accuracy': 0.8673020601272583,                                                                                                                                        
  'train/epoch': 0,                                                                                                                                                            
  'train/grad_norm/lm_head.kernel': 0.74609375,                                                                                                                                
  'train/grad_norm/model.norm.kernel': 0.00506591796875,                                                                                                                       
  'train/learning_rate': 4.973118120688014e-05,                                                                                                                                
  'train/loss': 0.5999301671981812,                                                                                                                                            
  'train/mean_accuracy': 0.880715012550354,                                                                                                                                    
  'train/mean_loss': 0.5317749381065369,                                                                                                                                       
  'train/perplexity': 1.8219926357269287,                                                                                                                                      
  'train/step': 271,                                                                                                                                                           
  'train/train/max_grad_norm': 1.609375,                                                                                                                                       
  'train/train/mean_grad_norm': 0.09228515625,                                                                                                                                 
  'train/visited_tokens': 2228224,                                                                                                                                             
  'train/z_loss': 0.0} 
```

**Comment by erfanzar** [comment]
@salrowili 
FLOPs are calculated automatically by JAX.

For logging during training, you have the flexibility to customize how you log or add logging hooks. By default, the logging method is set to `json`, but you can switch to `tqdm` by specifying it in the `TrainingArguments` under `progress_bar_type`.

If the current logging format doesn't suit your preferences, reverting to `tqdm` as the default might be a better option, as it provides a more intuitive and user-friendly progress display.

**Comment by salrowili** [comment]
You are right. However, i think it would be better to have the default progress bar type to tqdm because json will flood the terminal with loggings especially when you set the log step to be smaller.

I have re-run the code and fix an issue that the dataset in TPUv3-8 uses the cached dataset, not the updated one. 

TPUv3-8

`training process:  97%|▉| 252/260 [02:48<00:05,  1.55it/s, mlperf/execution_time=0.378, mlperf/flops=8.25e+12, mlperf/flops_per_sequence=4.07e+9, mlperf/flops_per_token=3.`

TPUv4-8

`training process:  98%|▉| 256/260 [04:37<00:04,  1.06s/it, mlperf/execution_time=0.773, mlperf/flops=1.15e+14, mlperf/flops_per_sequence=5.61e+10, mlperf/flops_per_token=5.4`

Also one thing related to jax jaxlib version. We need to fix the requirement of  the jax jaxlib version in the mean time to 0.4.35.

https://github.com/erfanzar/EasyDeL/blob/d13ecbbfd72713c5abf04744c1e10dd75e837ad7/pyproject.toml#L36-L37

This is because the function core.new_main was depreciated after jax >0.4.35 . This function is used by fjformer repo :

https://github.com/erfanzar/FJFormer/blob/f9940b097ef34ba5627ce6a259bfe629372899d8/fjformer/core/implicit_array.py#L452

However, fixing the new_main function would a better solution since the new jax 0.5.0 has been released.
https://github.com/jax-ml/jax/releases

**Comment by salrowili** [comment]
Hi @erfanzar .  Just to keep you updated, i think the new EasyDeL updates with Flax NNX have improved the memory efficiency. With 0.0.80 I usually get an OOM error if i try to run 8B model with 2048 max sequence and batch size of 8. I run these settings on TPUv3-8 with the new updated and no OOM error presented. I am carious to test the efficiency and speed of these new updated with TPUv4-8 once the issue that i mentioned in the previous reply get fixed.

**Comment by salrowili** [comment]
Hi @erfanzar ,
To update you in this issue, the speed now is getting better and i think its better than 0.0.80. You can close this issue. Just one note. I had to add this line for all codes in examples folder : 
`sys.setrecursionlimit(10000)`
